### PR TITLE
Future-facing Groovy 4 compatibility changes

### PR DESCRIPTION
**Pull request by DPUkyle**
Extracted from #20038 

These changes preserve existing Groovy 3 behavior, but should allow building and running Gradle with Groovy 4+

### CODE REVIEW

**Comment by DPUkyle** [comment]
@bot-gradle test this

File: subprojects/model-core/src/main/java/org/gradle/internal/reflect/PropertyAccessorType.java [hunk 1]
```diff
@@ -115,6 +117,10 @@ public static PropertyAccessorType of(Method method) {
         return null;
     }
 
+    private static boolean isGroovy3() {
+        return GroovySystem.getVersion().startsWith("3."); // FIXME total hack
```
Comment (jvandort) [inline subprojects/model-core/src/main/java/org/gradle/internal/reflect/PropertyAccessorType.java:121]: Do we want to make this less of a _total_ hack while we're here?

File: subprojects/build-init/src/main/java/org/gradle/buildinit/plugins/internal/JvmProjectInitDescriptor.java [hunk 2]
```diff
@@ -214,14 +212,20 @@ private void addDependencyConstraints(BuildScriptBuilder buildScriptBuilder) {
         buildScriptBuilder.implementationDependencyConstraint("Define dependency versions as constraints", "org.apache.commons:commons-text:" + commonsTextVersion);
 
         if (getLanguage() == Language.GROOVY) {
-            buildScriptBuilder.implementationDependencyConstraint(null, groovyModuleDependency("groovy-all", libraryVersionProvider.getVersion("groovy")));
+            buildScriptBuilder.implementationDependencyConstraint(null, getGroovyDependency(false));
```
Comment (jvandort) [inline subprojects/build-init/src/main/java/org/gradle/buildinit/plugins/internal/JvmProjectInitDescriptor.java:215]: How come we're not using `groovyModuleDependency` here anymore?

**Review by jvandort — APPROVED** [review]

File: subprojects/model-core/src/main/java/org/gradle/internal/reflect/PropertyAccessorType.java (hunk 1 above)
Comment (DPUkyle) [reply subprojects/model-core/src/main/java/org/gradle/internal/reflect/PropertyAccessorType.java:121]: Heh, yes.  The real reason for this is that our utility class `o.g.u.i.VersionNumber` is not available in the `model-core` subproject, and this small check doesn't warrant shuffling any classes or dependencies around.  I'll update the comment to reflect this.

File: subprojects/build-init/src/main/java/org/gradle/buildinit/plugins/internal/JvmProjectInitDescriptor.java (hunk 2 above)
Comment (DPUkyle) [reply subprojects/build-init/src/main/java/org/gradle/buildinit/plugins/internal/JvmProjectInitDescriptor.java:215]: Great point. Looking more closely at this, we can revert all these changes.  These were a workaround for https://issues.apache.org/jira/browse/GROOVY-10543.

**Comment by DPUkyle** [comment]
@bot-gradle test and merge

**Pull request merged**